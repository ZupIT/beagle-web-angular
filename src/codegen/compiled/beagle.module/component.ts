/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Type } from '@angular/core'
import {
  kebabToCamelCase,
  removeExtraIndentation,
  replaceToUnderline,
} from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  const componentName = replaceToUnderline(name)

  if (angularVersion >= 8) {
    return `'${componentName}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${componentName}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(
    components: Record<string, Type<any>>,
    angularVersion: number,
  ) {
  const queries = createQueries(components, angularVersion)

  // todo: legacy code. Remove the input "loadParams" with v2.0.
  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams', 'route', 'networkOptions', 'controllerId'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
