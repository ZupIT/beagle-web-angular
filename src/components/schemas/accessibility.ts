export interface Accessibility {
  accessible: boolean,
  accessibilityLabel?: string,
  isHeader?: boolean,
}
