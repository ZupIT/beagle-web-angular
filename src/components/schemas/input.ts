/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Properties as CSSProperties } from 'csstype'
import { InputHandler } from './input-handler'

export type InputType = 'DATE' | 'EMAIL' | 'PASSWORD' | 'NUMBER' | 'TEXT'

export interface BeagleInputInterface {
  value?: string,
  placeholder?: string,
  disabled?: boolean,
  readOnly?: boolean,
  type?: InputType,
  hidden?: boolean,
  styleId?: string,
  style?: CSSProperties,
  onChange?: InputHandler,
  onBlur?: InputHandler,
  onFocus?: InputHandler,
  error?: string,
  showError?: boolean,
}
