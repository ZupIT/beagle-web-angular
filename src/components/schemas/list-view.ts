/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { BeagleUIElement } from '@zup-it/beagle-web'
import { Properties as CSSProperties } from 'csstype'

export type Direction = 'VERTICAL' | 'HORIZONTAL'

export interface BeagleFutureListViewInterface {
  direction: Direction,
  dataSource: any[],
  iteratorName?: string,
  onInit?: () => void,
  onScrollEnd?: () => void,
  scrollEndThreshold?: number,
  template: BeagleUIElement,
  useParentScroll?: boolean,
  style?: CSSProperties,
}

export interface BeagleListViewInterface {
  direction: Direction,
}
