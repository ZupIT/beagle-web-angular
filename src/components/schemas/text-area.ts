/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Properties as CSSProperties } from 'csstype'

export type InputHandler = (event: { value: string }) => void

export interface TextAreaInterface {
  value?: string,
  label?: string,
  name?: string,
  onChange?: InputHandler,
  onFocus?: InputHandler,
  onBlur?: InputHandler,
  disabled?: boolean,
  readonly?: boolean,
  styleId?: string,
  style?: CSSProperties,
}
