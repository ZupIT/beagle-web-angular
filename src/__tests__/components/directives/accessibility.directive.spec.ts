/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { TestBed, async, ComponentFixture } from '@angular/core/testing'
import { buildBeagleTestModuleMetadata } from '../test-module-builder'
import { TestBeagleAccessibilityDirective } from '../mocks/accessibility.component'
import { Accessibility } from '../../../components/schemas/accessibility'

let component: TestBeagleAccessibilityDirective
let componentNative: HTMLElement
let fixture: ComponentFixture<TestBeagleAccessibilityDirective>

const beagleA11YttrsQuerySelector = 'p[data-test-id="p-element"]'
const ariaLabelText = 'Test Label'
const getAttrsAsArrayFromElement = (element: HTMLElement) =>  [...Array.from(element.attributes)]
const getOnlyAriaBeagleA11YttrsFn = a => a.name.includes('aria-')
const getOnlyRoleBeagleA11YttrsFn = a => (a.name === 'role')
const getBeagleA11YttrsFn = a => getOnlyAriaBeagleA11YttrsFn(a) || getOnlyRoleBeagleA11YttrsFn(a)
const getAttributeValue = (attrs: NamedNodeMap, name: string) => attrs.getNamedItem(name)?.value || ''
const buildAccessibility = 
  (accessible: boolean, accessibilityLabel: string, isHeader: boolean): Accessibility => 
    ({ accessible, accessibilityLabel, isHeader })

describe('BeagleA11YDirective', () => {
  beforeEach(async(() => {
    jest.autoMockOn()
    TestBed.configureTestingModule(
      buildBeagleTestModuleMetadata([TestBeagleAccessibilityDirective]))
      .compileComponents()

    fixture = TestBed.createComponent(TestBeagleAccessibilityDirective)
    component = fixture.componentInstance
    componentNative = fixture.debugElement.nativeElement
    fixture.detectChanges()
  }))

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  // eslint-disable-next-line max-len
  it('should have no "aria-" attributes or "role" attribute when no accessibility is provided or "accessibility.accessible" is false', () => {
    const p = componentNative.querySelector(beagleA11YttrsQuerySelector) as HTMLParagraphElement    
    let attrs = getAttrsAsArrayFromElement(p).filter(getBeagleA11YttrsFn)
    
    expect(attrs.length).toEqual(0)

    component.accessibility = buildAccessibility(false, ariaLabelText, true)
    fixture.detectChanges()

    attrs = getAttrsAsArrayFromElement(p).filter(getBeagleA11YttrsFn)

    expect(attrs.length).toEqual(0)
  })

  it('should set the attribute [role="heading"] when "accessible" and "isHeader" are true', () => {
    const p = componentNative.querySelector(beagleA11YttrsQuerySelector) as HTMLParagraphElement    
    component.accessibility = buildAccessibility(true, '', true)
    fixture.detectChanges()

    const attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)

    expect(attrs.length).toEqual(0)
    expect(getAttributeValue(p.attributes, 'role')).toEqual('heading')
  })

  // eslint-disable-next-line max-len
  it(`should set the attribute [aria-label="${ariaLabelText}"] when accessibilityLabel is setted`, () => {
    const p = componentNative.querySelector(beagleA11YttrsQuerySelector) as HTMLParagraphElement    
    component.accessibility = buildAccessibility(true, ariaLabelText, false)
    fixture.detectChanges()

    let attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)
    expect(attrs.length).toBeGreaterThan(0)

    attrs = getAttrsAsArrayFromElement(p).filter(getOnlyRoleBeagleA11YttrsFn)
    expect(attrs.length).toBe(0)

    expect(getAttributeValue(p.attributes, 'aria-label')).toEqual(ariaLabelText)
  })

  it(`should have role and [aria-label="${ariaLabelText}"] when everything is truthy`, () => {
    const p = componentNative.querySelector(beagleA11YttrsQuerySelector) as HTMLParagraphElement    
    component.accessibility = buildAccessibility(true, ariaLabelText, true)
    fixture.detectChanges()

    let attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)
    expect(attrs.length).toBeGreaterThan(0)

    attrs = getAttrsAsArrayFromElement(p).filter(getOnlyRoleBeagleA11YttrsFn)
    expect(attrs.length).toBeGreaterThan(0)

    expect(getAttributeValue(p.attributes, 'role')).toEqual('heading')
    expect(getAttributeValue(p.attributes, 'aria-label')).toEqual(ariaLabelText)
  })
})
