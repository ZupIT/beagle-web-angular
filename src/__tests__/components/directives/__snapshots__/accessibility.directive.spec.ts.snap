// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleA11YDirective should match snapshot 1`] = `
<test-beagle-accessibility-directive>
  <p
    data-test-id="p-element"
  />
</test-beagle-accessibility-directive>
`;
